observability/logging.go

```diff
@@ -0,0 +1,215 @@
+				Type:    grpclogrecordpb.GrpcLogRecord_Address_Type(binlogEntry.Peer.Type),
+				Address: binlogEntry.Peer.Address,
+				IpPort:  binlogEntry.Peer.IpPort,
+			}
+		}
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_HEADER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_HEADER
+		if binlogEntry.Peer != nil {
+			grpcLogRecord.PeerAddress = &grpclogrecordpb.GrpcLogRecord_Address{
+				Type:    grpclogrecordpb.GrpcLogRecord_Address_Type(binlogEntry.Peer.Type),
+				Address: binlogEntry.Peer.Address,
+				IpPort:  binlogEntry.Peer.IpPort,
+			}
+		}
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetServerHeader().Metadata)
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_REQUEST_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().Data
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().Data
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_HALF_CLOSE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_HALF_CLOSE
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_TRAILER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_TRAILER
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetTrailer().Metadata)
+		grpcLogRecord.StatusCode = binlogEntry.GetTrailer().StatusCode
+		grpcLogRecord.StatusMessage = binlogEntry.GetTrailer().StatusMessage
+		grpcLogRecord.StatusDetails = binlogEntry.GetTrailer().StatusDetails
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+		callEnded = true
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CANCEL:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_CANCEL
+		callEnded = true
+	default:
+		return fmt.Errorf("unknown event type: %v", binlogEntry.Type)
+	}
+	switch binlogEntry.Logger {
+	case binlogpb.GrpcLogEntry_LOGGER_CLIENT:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_CLIENT
+	case binlogpb.GrpcLogEntry_LOGGER_SERVER:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_SERVER
+	default:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_UNKNOWN
+	}
+	if callEnded {
+		c.removeEntry(binlogEntry.CallId)
+	}
+	// CloudLogging client doesn't return error on entry write. Entry writes
+	// don't mean the data will be uploaded immediately.
+	c.exporter.EmitGrpcLogRecord(grpcLogRecord)
```

Comment:
This expects that `c.exporter` didn't change during the life of this function call, but it seems there is no guarantee of this.